### ISSUE TITLE
[Feature] L3 Serving Runtime Design

### ISSUE THREAD

**Issue by superxf** [hw-native-sys/pypto-serving#18]
## Goal

Move the current serving execution path to a Simpler L3 structure while
preserving the public HTTP/OpenAI-compatible API and the existing serving
scheduler semantics.

The target split is:

- The outer serving engine owns request lifecycle, continuous batching, prefix
  cache metadata, KV block allocation, streaming output, and cancellation.
- The L3 runtime owns one scheduled step's execution DAG: chip tasks, host
  sub-worker tasks, tensor dependencies, and task completion.

This keeps the service boundary stable and moves device execution into the
hierarchical runtime model already used by Simpler.

## Current Structure

The current serving path is:

```text
FastAPI
  -> AsyncLLMEngine.add_request()
    -> Scheduler.schedule()
      -> WorkerProcess._execute_step()
        -> _batch_prefill() / _batch_decode()
          -> executor.run_prefill() / executor.run_decode()
            -> L2 worker, or one-shot L3 generate fast path
```

Important files:

- `python/core/server.py`: HTTP/OpenAI-compatible API.
- `python/core/async_engine.py`: async request loop and scheduler dispatch.
- `python/core/scheduler.py`: continuous batching, prefix cache lookup,
  preemption, KV block allocation.
- `python/core/serving_worker.py`: worker process that executes scheduled
  prefill/decode batches.
- `python/runtime/worker.py`: thin wrapper over `simpler.worker.Worker` for
  L2 and L3.
- `examples/model/qwen3_14b/runner/npu_runner.py`: Qwen3 NPU runner, including
  the existing one-shot `run_generate_l3()` path.

The current serving worker executes prefill first, then decode, by calling
Python executor methods directly. That means serving has an outer scheduler,
but the per-step execution is not yet represented as an L3 DAG.

## Target Structure

The desired serving path is:

```text
FastAPI
  -> AsyncLLMEngine.add_request()
    -> Scheduler.schedule()
      -> WorkerProcess._execute_step()
        -> L3ServingRuntime.run_step()
          -> Worker(level=3).run(orchestrator)
            -> orch.submit_next_level(prefill/decode chip callable)
            -> orch.submit_sub(sample / prepare / stop-check)
            -> drain
        -> StepOutput(new_tokens)
```

`AsyncLLMEngine` and `Scheduler` should not know Simpler details. They continue
to exchange `SchedulerOutput` and `StepOutput` with the worker process.

The new L3-specific logic should live behind a small runtime facade:

```text
python/core/l3_serving_runtime.py
  L3ServingRuntime
    init()
    close()
    run_step(scheduler_output) -> StepOutput
    run_prefill_step(...)
    run_decode_step(...)
```

`WorkerProcess` owns one `L3ServingRuntime` for its lifetime. It initializes it
after model loading and closes it when the worker process shuts down.

### Serving API Layer

Files: `python/core/server.py`, `python/cli/main.py`.

Responsibilities stay unchanged:

- Parse OpenAI-compatible completion/chat requests.
- Create `GenerateConfig`.
- Stream or aggregate token outputs.
- Start/stop `AsyncLLMEngine`.

No L3-specific code should be added here.

### Serving Engine Layer

Files: `python/core/async_engine.py`, `python/core/scheduler.py`,
`python/core/kv_cache.py`.

Responsibilities stay mostly unchanged:

- Tokenize prompts.
- Maintain request states.
- Run continuous batching.
- Decide which requests are prefill vs decode in each iteration.
- Allocate and release KV block ids.
- Track prefix cache metadata.
- Produce streaming token outputs.

The scheduler remains the owner of request lifecycle. L3 does not decide which
requests run or when a request is finished globally.

### Worker Process Layer

File: `python/core/serving_worker.py`.

`WorkerProcess` becomes a lifecycle shell around the L3 runtime:

- Load model and tokenizer-side metadata as today.
- Construct the executor.
- Register model artifacts with the executor.
- Build and initialize `L3ServingRuntime`.
- On each `WorkerCommand(type="step")`, call `l3_runtime.run_step()`.
- Return `StepOutput`.

This replaces `_batch_prefill()` and `_batch_decode()` as the primary NPU path.
The old methods can stay as a fallback while the L3 runtime is being rolled out.

### L3 Runtime Layer

New file: `python/core/l3_serving_runtime.py`.

Responsibilities:

- Own a long-lived `simpler.worker.Worker(level=3)`.
- Register chip callables before `worker.init()`.
- Register sub-worker callables before `worker.init()`.
- Own reusable shared-memory tensors for step inputs/outputs.
- Own child-memory device buffers for static weights and KV pages when possible.
- Build the per-step L3 orchestrator.
- Convert scheduler output into L3 task arguments.
- Return sampled token ids as `StepOutput`.

The L3 runtime should support two execution styles:

- `prefill` step: scheduled requests whose prompt is not fully computed.
- `decode` step: scheduled requests that already have at least one generated
  token or have completed prefill.

The scheduler may output both prefill and decode requests in the same iteration.
The first implementation may serialize these two groups inside one L3 run:

```text
L3 run:
  submit all prefill chip tasks
  submit sample tasks for completed-prefill rows
  submit all decode chip tasks
  submit sample tasks for decode rows
```

Later, if the underlying generated program supports it, prefill and decode can
be fused into a single compiled L3 program.

### L2 Chip Layer

The L2 chip layer remains Simpler-managed. L3 submits chip callables via
`orch.submit_next_level()`.

The L3 runtime should not reimplement chip dispatch. It should use the same
registered callable ids and `CallConfig` model used by Simpler's hierarchical
runtime.

## Why Not Reuse `run_generate_l3()` Directly

`run_generate_l3()` currently runs full generation inside one
`Worker(level=3).run()` call. That is good for one-shot CLI generation, but it
does not match serving requirements:

- Serving needs one token output at a time for streaming.
- Serving needs cancellation and abort handling.
- Serving needs continuous batching across many requests.
- Serving needs the scheduler to own prefix cache and KV block allocation.
- Serving may schedule prefill chunks and decode rows together.

Therefore serving should use L3 for a scheduled step, not for the whole request
lifecycle.

The existing `run_generate_l3()` is still valuable as a source of reusable
implementation pieces:

- Generated L3 artifact extraction.
- `Worker(level=3)` construction.
- Chip callable registration.
- Sub-worker registration.
- Static weight pre-upload using child-memory tensors.
- KV cache device pointer handling.
- Sampling and next-token preparation sub-worker logic.

### WorkerProcess

Add one optional member:

```python
self.l3_runtime: L3ServingRuntime | None = None
```

Initialization:

```python
self.l3_runtime = L3ServingRuntime(
    model_record=self.model_record,
    executor=self.executor,
    platform=self.config.platform,
    device_ids=[self.config.device_id],
)
self.l3_runtime.init()
```

Step execution:

```python
if self.l3_runtime is not None:
    return self.l3_runtime.run_step(scheduler_output)
return self._execute_step_legacy(scheduler_output)
```

### L3ServingRuntime

Suggested shape:

```python
class L3ServingRuntime:
    def __init__(self, model_record, executor, platform, device_ids):
        ...

    def init(self) -> None:
        ...

    def close(self) -> None:
        ...

    def run_step(self, scheduler_output) -> StepOutput:
        prefill = [sr for sr in scheduler_output.scheduled_requests if sr.is_prefill]
        decode = [sr for sr in scheduler_output.scheduled_requests if not sr.is_prefill]
        new_tokens = {}
        if prefill:
            new_tokens.update(self.run_prefill_step(prefill))
        if decode:
            new_tokens.update(self.run_decode_step(decode))
        return StepOutput(new_tokens=new_tokens)
```

The public contract should remain `StepOutput(new_tokens={request_id: token})`.
That keeps `AsyncLLMEngine._process_step_output()` unchanged.

### Inputs From Scheduler

Each `ScheduledRequest` provides:

- `request_id`
- `num_new_tokens`
- `num_computed_tokens`
- `is_prefill`
- `block_ids`
- request prompt tokens and generated tokens through `scheduled.request`

The L3 runtime must build:

- token id tensors
- embedding tensors
- position tensors
- sequence length tensors
- block table tensors
- slot mapping tensors

These are currently assembled in `WorkerProcess._batch_prefill()` and
`WorkerProcess._batch_decode()`. That code should move behind the L3 runtime,
with the output buffers converted to shared memory because L3 child processes
need to see them.

### Outputs To Scheduler

The L3 runtime returns:

```python
StepOutput(new_tokens={request_id: token_id})
```

The scheduler then:

- Increments `num_computed_tokens`.
- Caches completed prefix blocks.
- Appends new output token ids.
- Checks EOS and length stop conditions.
- Releases blocks for finished requests.

This logic should remain in `Scheduler.update_from_output()`.

## KV Cache Ownership

Keep a strict boundary:

- Main process scheduler owns KV metadata: block ids, prefix hashes, allocation,
  release, preemption, and cache hits.
- Worker/L3 runtime owns physical device buffers and per-step block tables.

Do not let the L3 orchestrator allocate serving KV blocks. It may allocate
temporary device buffers and child-memory mirrors, but request KV block ids must
come from the outer scheduler. Otherwise prefix cache and preemption will drift.

The L3 runtime receives block ids through `ScheduledRequest.block_ids` and
materializes the corresponding block table and slot mapping for chip tasks.

## Static Buffers And Reuse

Serving should avoid creating and destroying a `Worker(level=3)` for every
request or every step.

Long-lived state:

- L3 worker.
- Registered chip callables.
- Registered sub-worker callables.
- Compiled L3 metadata.
- Static model weights in shared memory.
- Optional child-memory device pointers for static weights.
- KV cache device buffers if their lifetime matches the worker.

Per-step state:

- Scheduled request view.
- Token tensors.
- Position tensors.
- Block table and slot mapping.
- Output logits or sampled-token buffers.
- Temporary task args.

The first implementation can be conservative and allocate per-step shared
memory tensors. Optimization can later move common buffers into reusable pools.

## Sub-Worker Tasks

Use SUB workers for host-side work that must run inside the L3 DAG:

- Sampling from logits.
- Preparing next decode input for a scheduled step.
- Writing sampled token ids into a shared result buffer.
- Lightweight stop checks that only need token ids.

Do not put request lifecycle decisions in SUB workers. Final request completion
still belongs to the outer scheduler.

### Phase 1: L3 Runtime Skeleton

- Add `python/core/l3_serving_runtime.py`.
- Add `L3ServingRuntime.init()/close()` with long-lived
  `Worker(level=3)`.
- Register placeholder sub-worker callables.
- Wire `WorkerProcess` to create/close the runtime when `executor_kwargs`
  requests L3 serving.
- Keep legacy `_batch_prefill()` and `_batch_decode()` fallback.

Acceptance:

- Unit tests still pass with default non-L3 serving.
- L3 serving can start and stop without running a request.

### Phase 2: Decode Step Through L3

- Implement `run_decode_step()` for scheduled decode rows.
- Reuse current decode input preparation from `_batch_decode()`.
- Submit decode chip callable via L3 orchestrator.
- Submit sampling as SUB worker.
- Return `StepOutput`.

Acceptance:

- A request with prefill already completed can generate one decode token.
- Streaming path receives the token through unchanged `AsyncLLMEngine` logic.

### Phase 3: Prefill Step Through L3

- Implement `run_prefill_step()` for prompt chunks.
- Reuse current prefill input preparation from `_batch_prefill()`.
- Submit prefill chip callable via L3 orchestrator.
- Only sample when the scheduled chunk completes the prompt.

Acceptance:

- Non-chunked prefill produces the first generated token.
- Chunked prefill returns no token until the final prompt chunk.

### Phase 4: Mixed Prefill And Decode

- Support scheduler outputs containing both prefill and decode requests.
- Serialize prefill and decode groups inside one `run_step()` initially.
- Later optimize by submitting a single combined DAG if generated programs
  support it.

Acceptance:

- Continuous batching works with one request in prefill and another in decode.

### Phase 5: Restore Serving Features

Re-enable and test:

- Prefix cache.
- Chunked prefill.
- Preemption.
- Batch size greater than one.
- Stop strings at API level.
- EOS and length stopping.
- Streaming and non-streaming completions.

### Phase 6: Performance Work

- Keep L3 worker alive for process lifetime.
- Pre-upload static weights once per worker lifetime.
- Reuse shared-memory buffers by max serving batch shape.
- Keep KV cache device buffers resident where possible.
- Add timing around scheduler, L3 orchestrator, chip task, sub-worker, and
  result processing.

## Compatibility Strategy

Use a feature flag during migration:

```json
{
  "npu": {
    "l3": true,
    "l3_serving": true
  }
}
```

Interpretation:

- `l3=true`: compile/register L3-capable artifacts.
- `l3_serving=true`: route serving steps through `L3ServingRuntime`.

This keeps the existing one-shot `--l3` path and the new serving L3 path
separate while the serving path matures.

### Two Schedulers Can Conflict

The outer serving scheduler and Simpler L3 scheduler operate at different
levels. The outer scheduler must own request lifecycle; the L3 scheduler must
own only per-step task dependencies. Avoid moving request admission, prefix
cache lookup, or KV block allocation into L3.

### Worker Lifetime And Registration

Simpler L3 requires chip and sub callables to be registered before
`worker.init()`. Serving initialization must compile/extract all callables
before starting the L3 worker.

### Shared Memory Requirements

L3 child processes see tensors through shared memory or child-memory device
pointers. Any tensor passed to L3 child/sub tasks must be explicitly placed in
shared memory or converted to a valid `ContinuousTensor`.

### Batch Shape Drift

Current one-shot L3 generation supports batch size 1. Serving uses dynamic
batching. The first L3 serving milestone should either enforce a known batch
tile or explicitly pad to the compiled tile shape.

### KV Sync Cost

Copying full KV state back to host after every step defeats serving
performance. The design should aim for worker-resident KV pages and only sync
metadata to the host scheduler. Host-visible KV materialization should be a
debug/fallback path.
